File: backend/app/models/openrouter.py
from typing import Dict, List, Optional, Union, Any
from pydantic import BaseModel, Field


class Message(BaseModel):
    """Model for a chat message in OpenRouter format"""
    role: str
    content: str


class CompletionRequest(BaseModel):
    """Model for the OpenRouter API request payload for completions"""
    model: str
    messages: List[Message]
    temperature: Optional[float] = 0.7
    top_p: Optional[float] = 0.95
    max_tokens: Optional[int] = 1024
    stream: Optional[bool] = False
    stop: Optional[Union[str, List[str]]] = None
    frequency_penalty: Optional[float] = 0
    presence_penalty: Optional[float] = 0


class Choice(BaseModel):
    """Model for a completion choice returned by OpenRouter"""
    index: int
    message: Message
    finish_reason: Optional[str] = None


class Usage(BaseModel):
    """Model for token usage information"""
    prompt_tokens: int
    completion_tokens: int
    total_tokens: int


class CompletionResponse(BaseModel):
    """Model for the OpenRouter API response for completions"""
    id: str
    object: str = "chat.completion"
    created: int
    model: str
    choices: List[Choice]
    usage: Usage


class AvailableModel(BaseModel):
    """Model for available OpenRouter model information"""
    id: str
    name: str
    description: Optional[str] = None
    pricing: Optional[Dict[str, Any]] = None
    context_length: Optional[int] = None
    top_provider: Optional[str] = None


class ModelsResponse(BaseModel):
    """Model for the OpenRouter API response for available models"""
    data: List[AvailableModel]


class APIKeyResponse(BaseModel):
    """Model for API key management responses"""
    key: str
    created_at: str


class APIKeyListResponse(BaseModel):
    """Model for a list of API keys"""
    keys: List[APIKeyResponse] 

File: frontend/src/services/api/openrouter.ts
import { api } from './config';

export interface Message {
  role: string;
  content: string;
}

export interface CompletionRequest {
  model: string;
  messages: Message[];
  temperature?: number;
  top_p?: number;
  max_tokens?: number;
  stream?: boolean;
  stop?: string | string[];
  frequency_penalty?: number;
  presence_penalty?: number;
}

export interface Choice {
  index: number;
  message: Message;
  finish_reason?: string;
}

export interface Usage {
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
}

export interface CompletionResponse {
  id: string;
  object: string;
  created: number;
  model: string;
  choices: Choice[];
  usage: Usage;
}

export interface AvailableModel {
  id: string;
  name: string;
  description?: string;
  pricing?: Record<string, any>;
  context_length?: number;
  top_provider?: string;
}

export interface ModelsResponse {
  data: AvailableModel[];
}

export interface APIKey {
  key: string;
  created_at: string;
}

export interface APIKeyListResponse {
  keys: APIKey[];
}

export interface ActiveKeyInfo {
  active: boolean;
  key?: string;
  message?: string;
}

class OpenRouterService {
  async getModels(): Promise<ModelsResponse> {
    const response = await api.get('/openrouter/models');
    return response.data;
  }

  async createCompletion(request: CompletionRequest): Promise<CompletionResponse> {
    const response = await api.post('/openrouter/chat/completions', request);
    return response.data;
  }

  async listKeys(): Promise<APIKeyListResponse> {
    const response = await api.get('/openrouter/keys');
    return response.data;
  }

  async addKey(key: string): Promise<APIKey> {
    const response = await api.post('/openrouter/keys', { key });
    return response.data;
  }

  async deleteKey(key: string): Promise<void> {
    await api.delete(`/openrouter/keys/${key}`);
  }

  async setActiveKey(key: string): Promise<{ message: string }> {
    const response = await api.post('/openrouter/keys/active', { key });
    return response.data;
  }

  async getActiveKey(): Promise<ActiveKeyInfo> {
    const response = await api.get('/openrouter/keys/active');
    return response.data;
  }
}

export const openRouterService = new OpenRouterService(); 